### PR TITLE
add members_info state query and API wrapper

### PR DESCRIPTION
**Pull request by illotum**
## Proposed Changes

Adds state query and API to return leader's cluster view.

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in related repositories

### CODE REVIEW

**Comment by illotum** [comment]
Opening for review. 

@kjnilsson what if we called it `cluster_info` instead? Considering the return type and internal representation, it feels a little more intuitive.

Second question, should I send zeroes instead of no values for the ease of matching? This mainly concerns the corner case of a local query to non-leader.

**Comment by michaelklishin** [comment]
@illotum `cluster_info` is not a bad name but note that a cluster status is (or at least can be) more than the status of its members. `members_info` is more specific in that sense.

File: src/ra_server_proc.erl [hunk 1]
```diff
@@ -1503,13 +1505,7 @@ follower_leader_change(Old, #state{pending_commands = Pending,
...
     case node() of
         Node -> ok;
-        _ ->
-            case aten:register(Node) of
-                ignore ->
-                    ok;
-                Res ->
-                    Res
-            end
+        _ -> aten:register(Node)
```
Comment (michaelklishin) [inline src/ra_server_proc.erl:1508]: This is not a direct equivalent (maybe this does not matter). In the original version a value of `ignore` returned by Aten would be returned as `ok` by Ra. Now that's not the case, is there a reason for that?

Comment (illotum) [reply src/ra_server_proc.erl:1508]: Dialyze mostly! `aten:register` returns only `ok`.

**Comment by illotum** [comment]
Added a commit to extract membership from `voter_status` key. I'm uncertain if this is better than full voter_status, but clearly it is the main value of that entry. I can think of a situation where `target` may be valuable, so happy to revert.

**Review by michaelklishin — APPROVED** [review]

File: src/ra_server_proc.erl [hunk 2]
```diff
@@ -1555,6 +1551,51 @@ do_state_query(voters, #{cluster := Cluster}) ->
...
+    maps:map(fun(Id, Peer) ->
+                 case {Id, Peer} of
+                     {Self, Peer = #{voter_status := _}} ->
+                         %% For completeness sake, preserve `target`
+                         %% of once promoted leader.
+                         #{next_index => CI+1,
+                          match_index => CI,
+                          query_index => QI,
+                          status => normal,
+                          voter_status => Peer#{membership => Membership}};
```
Comment (kjnilsson) [inline src/ra_server_proc.erl:1566]: Sould this be the voter status map instead of the full `Peer`?

File: src/ra_server_proc.erl (hunk 1 above)
Comment (kjnilsson) [reply src/ra_server_proc.erl:1508]: I think you have an old aten version, see https://github.com/rabbitmq/aten/blob/main/src/aten.erl#L18

**Review by kjnilsson — CHANGES_REQUESTED** [review]
Couple of smaller changes needed but looks good.

Also would you mind running an indentation pass on the new code in ra_server_proc. It's all a bit wonky :)

**Comment by kjnilsson** [comment]
> Opening for review.
> 
> @kjnilsson what if we called it `cluster_info` instead? Considering the return type and internal representation, it feels a little more intuitive.


I think the link to `ra:members/1` works in terms of it's name so let's keep it as is. 

> 
> Second question, should I send zeroes instead of no values for the ease of matching? This mainly concerns the corner case of a local query to non-leader. On the flip side, I fear sending anything invalid will lead to user confusion.

I think what we're hitting here is the fact that this query makes less sense as a local query as really you want to issue it against the leader. I suggest when used as a local query we only return infos that make sense to return, e.g. the a=`voter_status`. Alt. we always make this query a leader query. Happy to consider either way.

File: src/ra_server_proc.erl (hunk 2 above)
Comment (illotum) [reply src/ra_server_proc.erl:1566]: That's what I get for working late night!

**Comment by illotum** [comment]
I'm leaning towards returning _something_. There are cases where one may want to investigate follower's view of the cluster, however limited that is.

**Comment by michaelklishin** [comment]
In the context of RabbitMQ's needs of determining if a node hosts any "catching up" (non-voter) replicas, either can work but a leader query makes more sense to me.

**Review by michaelklishin — APPROVED** [review]

**Pull request merged**